Make OptimizationResult read-only (all parameters of the constructor become mandatory)

t-imamichi opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add it in the CHANGELOG file under Unreleased section.
⚠️ If your pull request fixes an open issue, please link to the issue.

✅ I have added the tests to cover my changes.
✅ I have updated the documentation accordingly.
✅ I have read the CONTRIBUTING document.
-->

### Summary

This PR makes `OptimizationResult` read-only.
I found that most of existing codes do not modify the internal data of `OptimizationResult`.
So, I removed all public setter of `OptimizationResult`. It makes more robust because users cannot modify it with public methods.
It also helps mypy and pylint to detect  wrong modification of results.

Related #1107, #1122

Minor changes:
- Renames `OptimizationResults.results` with `OptimizationResults.raw_results`
- Converts the argument `x` of the constructor of `OptimizationResults` into ndarray to set as `OptimizationResult._x`.
- Adds the main docstring of `OptimizationResults` to state the order of variables.
- Fixes a bug of 'MinimumEigenOptimizer'. See below.
- Fixes pylint

Discussion about `OptimizationResult.raw_results` continues at another issue #1122.


### Details and comments

I slightly modified the converters to avoid direct access of the internals of `OptimziationResult`.
See https://github.com/Qiskit/qiskit-aqua/pull/1131#discussion_r459457660 for details.

I also found a bug of 'MinimumEigenOptimizer'.
`MinimumEigenOptimizer.solve` generates `str`, e.g. "110", while we expect `List[float]`, e.g. [1, 1, 0].
https://github.com/Qiskit/qiskit-aqua/blob/146bff86d486dfd73a871e7f3c000015bc872581/qiskit/optimization/algorithms/minimum_eigen_optimizer.py#L164
Due to this bug,  the following unit test fails.
https://github.com/Qiskit/qiskit-aqua/blob/146bff86d486dfd73a871e7f3c000015bc872581/test/optimization/test_grover_optimizer.py#L43
I fixed it in this PR.

I also change the name `OptimizationResults.results` into `OptimizationResults.raw_results`.

adekusar-drl commented on qiskit/optimization/algorithms/optimization_algorithm.py:
```diff
@@ -108,8 +109,8 @@ class OptimizationResult:
 
-    def __init__(self, x: Optional[Any] = None, fval: Optional[Any] = None,
-                 results: Optional[Any] = None,
+    def __init__(self, x: Union[List[float], np.ndarray], fval: float,
+                 results: Any,
```
I guess we can keep `results` as optional?

t-imamichi replied:
It's possible to keep it as optional, but all algorithms pass something to `results` now. So, I removed `Optional`. Do you want keep it as optional?

adekusar-drl replied:
I would keep it optional but I'm with either solution.

t-imamichi replied:
OK. Then, I will make it optional.

adekusar-drl commented on qiskit/optimization/converters/integer_to_binary.py:
```diff
@@ -215,14 +215,12 @@ def decode(self, result: OptimizationResult) -> OptimizationResult:
-        vals = result.x
-        new_vals = self._decode_var(vals)
-        result.x = new_vals  # type: ignore
+        result._val = self._decode_var(result.x)  # type: ignore
         return result
```
Any chance we can avoid accessing `_val` directly?

t-imamichi replied:
It is difficult because this method receives an object of various `OptimizationResult` children such as ADMMOptimizationResult, etc. and interprets the values of binary variables into integer variables.
So, it is hard to use the constructor of the children. But, I can think of a helper function to do it. Let me think...

t-imamichi replied:
If we make a following copy method, we can avoid accessing `_val`. It may be better than a public setter of `self.x`. What do you think? 
```python
class OptimizationResult:
...
    def copy(self, x: Union[List[float], np.ndarray]) -> OptimizationResult:
        new_result = copy.deepcopy(self)
        new_result._val = x
        return new_result
```

t-imamichi replied:
If no algorithm extends `OptimizationResult`, it is easy. We just call the constructor with updated values.

adekusar-drl replied:
This is a very good question. Do we need sub-classes of `OptimizationResult`? Frankly I don't really get the benefits of sub-classing results.
Anyway, if we can't come up with a better solution we can keep direct access, it is just python, everything is public no matter how hard we try.

adekusar-drl replied:
Oh, I'd say adding `copy` we make it more complicated. I prefer to have direct access :)

adekusar-drl replied:
> 
> 
> If no algorithm extends `OptimizationResult`, it is easy. We just call the constructor with updated values.

That's the question. What's the reason for sub-classes? I don't see serious reasons for this.

a-matsuo replied:
It's because different algorithms output `OptimizationResults` with  different additional informations. For example, `ADMMOptimizationResults` has `state` property, and `MinimumEigenOptimizerResult` has `samples` property etc. We can create an `OptimizationResults` class with lots of properties that most of those property will be empty most of the times. I'm not sure which way is better, crating sub-classes or creating an `OptimizationResult` class with lots of properties.

adekusar-drl replied:
@a-matsuo Matsuo-san, In case of ADMM we have a sub-class, but we fill in both `results` and `state` with the absolutely same value. So, in case of ADMM I don't see why we have this, except that Stefan told me to do so :)
I don't think we need many-many properties defined in the `OptimizationResult` class. I believe we only need `results` as a generic property where we may keep any value we want. And as a conclusion I don't see a reason why we need sub-classes, I think we may have just a basic `OptimizationResult` for all purposes.

woodsp-ibm replied:
In Aqua we used to have a dictionary return for the algorithms which we are trying to go away from. This is all very well and extensible but hard to know what the keys are and the IDE can't help you the same way it can do with getters which are much easy to see/discover. Hence we went over to result classes. The intent was to keep info the user would be interested in visible this way. And so for example we have a commonn min eigen result which VQE extends and adds unique fields it has. Same for QPE etc. The more it goes to some generic catch all the less visible things can become. Now in chemistry result there is an algorithm result field that contains whatever the raw result was of Aqua algorithm used - maybe this akin to the results field you are talking about.

adekusar-drl replied:
So you have a long story of using result classes in Aqua. In the optimization stack we have a similar issue, the difference to chemistry, as I got it, that we have two mandatory(if everything goes well) properties, a float and an array of floats(`fval` is a cost function value and `x` is a solution vector). The third parameter(`results`) is up to an optimizer. It can be everything. Most of time, I guess, those two properties are the things a user may need(may be I'm wrong here). That's why I suggest to keep `fval` and `x` as mandatory properties with certain type hints and have `results` as an optional and unspecified. If somebody wants and knows the concrete result class it is still possible to add a type hint to resolve `results` type. It is not a great solution, not cast in stone but python is a serious OO language. 
In overall, if you think we should keep as many type hints as possible at a cost of some overhead - fine with me, I like to know what I have as an output.

woodsp-ibm replied:
In my mind it would be say whether an SLSPQOptimzer or ADMMOptimizer would want/need to subclass the generic optimizer result to add unique fields that may be of interest/use to a user when you know its the specific optimizer you used, We did this via subclassing in Aqua core algorithms so VQE has a result, class and so does QPE etc, each extending a more generic result. For chemistry application module we just put the 'raw' algorithm result - ie base class from which all Aqua core algos derive https://github.com/Qiskit/qiskit-aqua/blob/146bff86d486dfd73a871e7f3c000015bc872581/qiskit/chemistry/core/chemistry_operator.py#L111
For me its less about typehints and stuff, but more about the user being able to find/discover result parts which can be more easily seen through some public API with setters and is more opaque with a dictionary. What the right balance is for Optimization in making say specific getters versus dropping a dictionary or other complex result object into some generic results field I cannot say. I am merely noting some of the thinking around what we did so far for Aqua (and still are in the process of conversion to in some cases)

adekusar-drl replied:
I see, thanks for sharing this. In the optimization stack we don't have a dictionary. But anyway, since you've already implemented a sub-classing pattern, let's keep it in optimization also.

t-imamichi replied:
If we keep subclass pattern, we need to choose one of the following options.
1. Allow a direct access `OptimizationResult._val` to only `IntegerToBinaryConverter.decode` as an exceptional case.
2. Make a public setter of `OptimizationResult.x`
3. Make a copy method as I wrote https://github.com/Qiskit/qiskit-aqua/pull/1131#discussion_r458893385

I think `OptimizationRestuls._results` is regarded as the raw results of each algorithm. So, how about changing the name of the class variable as `OptimizationResults._raw_results` and the getter as `OptimizationResults.raw_results` to be more clear? If we prohibit sub-classes of OptimizationResult and make algorithms use `_raw_data` to embed algorithm specific data or `AlgorithmResult` as @woodsp-ibm mentioned, we can easily interpret the data without direct access of `_val`. https://github.com/Qiskit/qiskit-aqua/blob/146bff86d486dfd73a871e7f3c000015bc872581/qiskit/aqua/algorithms/algorithm_result.py#L25

adekusar-drl replied:
I'm in favor of 1. in this case.
Let's keep sub-classing.

t-imamichi replied:
I notice a technique to avoid the direct access!
For example, ADMM currently generates a result as follows.
```python
result = ADMMOptimizationResult(solution, objective_value, self._state)
# convert back integer to binary
result = cast(ADMMOptimizationResult, int2bin.decode(result)) # `int2bin.decode` needs to access `_val` because `result is ADMMOptimizationResult.
```
https://github.com/Qiskit/qiskit-aqua/blob/146bff86d486dfd73a871e7f3c000015bc872581/qiskit/optimization/algorithms/admm_optimizer.py#L361

By changing the logic, `int2bin` interprets the base class `OptimizationResult` by calling the constructor. We can also delete `cast` too.
```
# convert back integer to binary
base_result = OptimizationResult(solution, objective_value)
base_result = int2bin.decode(base_result)  # `int2bin.decode` can make a new result by calling the constructor of OptimizationResult
result = ADMMOptimizationResult(base_result.x, base_result.fval, self._state)
```

It works for MinimumEigenOptimizer too. I made a branch to try this technique.
commit: https://github.com/t-imamichi/qiskit-aqua/commit/2ae7f5faf8e921f9e41bdcc7fa0834193dc8f171
branch: https://github.com/t-imamichi/qiskit-aqua/tree/opt_result2

What do you think?

t-imamichi replied:
If you are OK. I merge the branch to this pullreq to continue the discussion.

molar-volume replied:
@t-imamichi maybe you can also merge #1122 to this branch to resolve the conflicts,
or if you want, I can merge this branch to #1122

adekusar-drl replied:
Sorry, I missed this comment. Looks fine to me.

t-imamichi replied:
I adjust the code with the merged PR #1122

adekusar-drl commented on test/optimization/test_converters.py:
```diff
@@ -406,7 +406,7 @@ def test_binary_to_integer(self):
         conv = IntegerToBinary()
         _ = conv.encode(op)
-        result = OptimizationResult(x=[0, 1, 1, 1, 1], fval=17)
+        result = OptimizationResult(x=[0, 1, 1, 1, 1], fval=17, results=0)
         new_result = conv.decode(result)
         self.assertListEqual(new_result.x, [0, 1, 5])
         self.assertEqual(new_result.fval, 17)
```
Could it be better to use `self.assertAlmostEqual` or `np.testing.assert_almost_equal` for floats? Just in case. Same for all float testing.

t-imamichi replied:
In this test, the values should be integers (casted to float), so both assertAlmostEqual and assertEqual work.  But, the type hint is float, so it might be good to use the almost equal version as you said.

t-imamichi commented:
I adjusted my pullreq merging master coping with the new attribute `Optimization.variables`.

woodsp-ibm commented on qiskit/optimization/algorithms/optimization_algorithm.py:
```diff
@@ -217,11 +173,29 @@ def status(self) -> OptimizationResultStatus:
+
+    @property
+    def variables_dict(self) -> Dict[str, int]:
+        """Returns the pairs of variable names and values under optimization.
```
Would this be better saying 'returns the optimal value a dictionary of the variables name and corresponding value' Saying 'pairs' makes me think too much of Tuple.

t-imamichi replied:
Agree. I change it.

woodsp-ibm commented on qiskit/optimization/algorithms/optimization_algorithm.py:
```diff
@@ -131,82 +132,37 @@ def __getitem__(self, item: Union[int, str]):
-
     @property
-    def results(self) -> Any:
+    def raw_results(self) -> Optional[Any]:
```
I think this can just be 'Any' since Any can include any type, which includes NoneType (the type of None)

t-imamichi replied:
Thanks. I will fix it.

t-imamichi commented on qiskit/optimization/algorithms/optimization_algorithm.py:
```diff
@@ -101,28 +103,27 @@ class OptimizationResult:
 
     def __repr__(self):
-        self._x = self._x if self._x is not None else []
         return 'optimal variables: [{}]\noptimal function value: {}\nstatus: {}' \
```
It makes sense. I will do it.

t-imamichi commented on qiskit/optimization/algorithms/optimization_algorithm.py:
```diff
@@ -131,82 +132,37 @@ def __getitem__(self, item: Union[int, str]):
             return self._variables_dict[item]
-        raise QiskitOptimizationError("Integer or string parameter required, instead "
-                                      + type(item).__name__ + " provided.")
+        raise QiskitOptimizationError(
```
Yes, I will do it.

t-imamichi commented on qiskit/optimization/algorithms/optimization_algorithm.py:
```diff
@@ -131,82 +132,37 @@ def __getitem__(self, item: Union[int, str]):
             return self._x[item]
```
I wrote the docstring and felt that the implementation is a bit tricky. I think it would be simpler to remove `__getitem__` to keep `OptimizationResult` simpler. Users can access the value of variables by `OptimizationResult.x[i]` if `i` is an integer and `OptimizationResult.variables_dict[name]` is `name` is a string. I remember that @adekusar-drl suggested this method. Do you want to keep `__getitem__`?

adekusar-drl replied:
@t-imamichi I think get item is a very convenient method. If there's no obvious drawback of difficulties, it's better to have it.

woodsp-ibm replied:
I have no issue with keeping method it does seem useful in some ways if you know about this behavior. Hence above when I gave the examples result[0] and result['A'] and referred to main docstring I was more thinking of some description in the class docstring that described this behavior rather than a docstring around the params etc on the getitem method. I do not know how the latter comes out, if at all, in the API docs. Its fine to have that on the method too but my original thought was more to document it in place where the user would more come across this. I did not pull this branch to check the html outcome as it stands.

t-imamichi replied:
OK. I keep `__getitem__`. I will add an example of the main docstrings of `OptimizationResult`. How about implementing `__len__`? It might be convenient to use with `__getitem__` with index.

woodsp-ibm replied:
Can I step back and ask some question. The optimal value is a list - how do I know the order in this list? Is it a one to one correspondence to the index of the variable that the quadratic program gives me. It seems the values are stored separately from the variables rather than being some field with the variables. Now the variables seems to be an Optional param for the result which means I need to know how the optimization result was built by the particular optimizer which may or may not provide the variables to the result. If it does not then the dictionary and names list I can get back are empty. It seems I can get back the optimal value as a list, or directly an item from that off the result via the get item, and if it were given variables then via a variable name on get item or a dictionary. There seem to be a few choices here, which maybe is not all bad, but then if variables are not provided I am left with the basic by index on the values only. If that is all I can rely on, knowing say the index from the Quadratic program variable to access the result why would I code it any other way? If I try to use the dictionary or access by name it will fail if variables were not supplied to the result so my code would have to be written that tried that and then fell back to by index. Given that's extra work, and the value is unclear to me since it has to cater to the basic case I guess I am asking about the overall usability of result here since things are being changed around it. Could variables me made mandatory to make the overall result more consistent in its behavior? If not I guess I am wondering why we have all that extra stuff based on names when I most likely am forced to really code by index to have it work everywhere.

adekusar-drl replied:
1. Variables/solution values must be indexed in the same way. Originally variables are stored as a list in `QuadraticProgram`, so in `OptimizationResult` solution values must be in the same order if you want access them by index.
2. I thought we decided to have variables (or at least variable names) as a mandatory parameter in  `OptimizationResult`. It does not cost a thing, since we already have a list of variables in `QudraticProgram`.
3. `__getitem__` is for convenience only, sometimes when you code a problem you want to solve, you remember variable names, but don't remember which of them was added first or second. It is similar to the behavior we've already introduced: `qp.objective.linear[index]`. Yes, a different story but in the code looks similar.
4. For `__len__` I don't see a real need at least now.

woodsp-ibm replied:
For 1. above the responsibility of the ordering is the optimizer producing the result, right? I did not see any definition of ordering hence I assumed they were the same as the QP as any other order did not seem to make sense.

For 2. certainly having it as mandatory would make the result more consistent/user friendly. Without it being mandatory you just can't rely on it at all, hence my comments.

For 3. get item by name (if you don't know the index) only works when it has the variables (Optional at present). There is also the dictionary return of key/values, when variables is provided, so I can do result['A'] or result.variables_dict['A'], the latter using the dictionary the former the result directly. Maybe its useful to have the whole dict be passed back as an option - either way access by name requires the variables. Part of my question around ordering was if I needed the index I guessed I could look it up in the QP, but that assumed the index order was the same which I did not see explicitly stated anywhere.

t-imamichi replied:
For 1. Both optimizers and converters are responsible for the ordering of the variables of the input QP. But, it is not explicitly stated. For example, `IntegerToBinary` and `IntegerToBinary` change come variables, so we need to take care of the ordering of variables of `encode` and `decode` (`convert` and `interpret`, new names).

It might be better to explicitly state the ordering of variables of `OptimizationResult` in `OptimizationAlgorithm` and `QuadraticProgramConverter` of #1061.

For 2. It is OK to make `variables` mandatory. It makes the code simpler.

t-imamichi replied:
I updated the docstring to add examples. Please take a look at it.

adekusar-drl replied:
I agree with @t-imamichi. The initial order of the variables comes from `QuadraticProgram`, after that it must be maintained in the rest code. Not great that it is not explicitly stated though.

t-imamichi replied:
I try to state it in the main docstring.

t-imamichi replied:
I updated the main docstring with examples.

adekusar-drl commented on qiskit/optimization/algorithms/optimization_algorithm.py:
```diff
@@ -217,11 +207,29 @@ def status(self) -> OptimizationResultStatus:
-        self._status = status
+        return self._variables
+
+    @property
+    def variables_dict(self) -> Dict[str, int]:
+        """Returns the optimal value as a dictionary of the variable name and corresponding value.
```
Likely `variables_dict` should return `Dict[str, float]` ?

t-imamichi replied:
Yes, you are right.

woodsp-ibm commented on qiskit/optimization/algorithms/optimization_algorithm.py:
```diff
@@ -101,112 +103,121 @@ class OptimizationResult:
+                 variables: List[Variable],
+                 raw_results: Optional[Any] = None,
+                 status: OptimizationResultStatus = OptimizationResultStatus.SUCCESS) -> None:
+        self._x = x  # pylint: disable=invalid-name
```
Minor comment: Since x can be a list[float] or an ndarray would it make sense to make it an ndarray in the case of a list so as it can from thereafter always be treated as an ndarray - e.g. return type of the setter.

t-imamichi replied:
Yes, unifying type into ndarray makes sense. I will change all `self.assertListEqual` into `np.testing.assert_almost_equal` accordingly.

t-imamichi replied:
I have done the change.

woodsp-ibm commented on qiskit/optimization/algorithms/optimization_algorithm.py:
```diff
@@ -101,112 +103,121 @@ class OptimizationResult:
-        self._variables = variables if variables is not None else []
+    def __init__(self, x: Union[List[float], np.ndarray], fval: float,
+                 variables: List[Variable],
+                 raw_results: Optional[Any] = None,
```
Should there be some structure to raw_results, as in rather than `Any` would `Dict[str, Any]` be preferable? I see this for instance the GroverOptimizer passes in a dictionary there. Not sure for other cases. At least the string field names give some clue to the content which might not otherwise be there if its Any directly.

adekusar-drl replied:
I'd suggest to keep it as is now and return to `raw_result` in #1120, if you don't mind.

t-imamichi replied:
I'd like to keep `raw_results` just `Any` in this PR and discuss details of raw_results in #1120 too. Is it OK?

woodsp-ibm replied:
Sure, was hoping that the Dict would cover things, but we can leave it as-is and visit it as part of that issue,

t-imamichi commented:
I have changed the followings:
- change the type of the internal of `OptimizationResult._x` from `Union[List[float], ndarray]` to `ndarray`
- added an example and a note about the order of variables in the main docstring
- fix pylint warnings

t-imamichi commented:
Travis seems to fail with a code not related to this PR.
> $ make lint
> pylint -rn --ignore=gauopen qiskit/aqua qiskit/chemistry qiskit/finance qiskit/ml qiskit/optimization test tools
> ************* Module aqua.components.neural_networks._pytorch_discriminator_net
> qiskit/aqua/components/neural_networks/_pytorch_discriminator_net.py:32:0: W0223: Method '_forward_unimplemented' is abstract in class 'Module' but is not overridden (abstract-method)

t-imamichi commented:
Torch issue is fixed at #1146.

t-imamichi commented:
The failed test it not related to this PR...
```
test.aqua.operators.test_pauli_expectation.TestPauliExpectation.test_pauli_expect_single
----------------------------------------------------------------------------------------
Captured traceback:
~~~~~~~~~~~~~~~~~~~
    Traceback (most recent call last):
      File "/home/travis/build/Qiskit/qiskit-aqua/test/aqua/operators/test_pauli_expectation.py", line 63, in test_pauli_expect_single
    self.assertAlmostEqual(sampled.eval(), matmulmean, delta=.1)
      File "/opt/python/3.7.1/lib/python3.7/unittest/case.py", line 893, in assertAlmostEqual
    raise self.failureException(msg)
    AssertionError: (-0.10546874999999996+0j) != (-2.198489195831717e-32-1.0146536357569526e-17j) within 0.1 delta (0.10546874999999996 difference)
```

woodsp-ibm commented:
That test passed under 3.8 but failed under 3.7. There are still one or two 'troublesome' tests that sometimes fail. I just restarted the job for 3.7 so lets see.

t-imamichi commented:
I hope the new build passes. Because the failed test is related to expectation value, it would be more robust if we can fix the random seed of the tests.

woodsp-ibm commented:
I have not looked at that specific test - in general seeds are set, but sometimes we still see an issue even when seeds are set. It passed after restarting. But yes if any test that seems to randomly fail can be improved we need to do it.

t-imamichi commented:
I see. I will also look at the test.

woodsp-ibm commented:
@t-imamichi I looked at the test and I can see its not using seeds for when its running on the backend. I will change what its doing to do that.

woodsp-ibm commented on qiskit/optimization/algorithms/optimization_algorithm.py:
```diff
@@ -101,112 +103,121 @@ class OptimizationResult:
+        {'x1': 0.0, 'x2': 1.0, 'x3': 0.0}
+
+    Note:
+        The order of variables are supposed to be equal to that of the problem solved by
```
```suggestion
        The order of variables should be equal to that of the problem solved by
```

woodsp-ibm commented:
@t-imamichi I think this is now done right? There is not any more work that needs to be done for this specific PR? I made one small change to the docs from supposed to be, to 'should be' - maybe it should even be stronger and say 'must be' for things to work right overall though I am not sure.

t-imamichi commented:
Thank you for your comments. I updated the docstring. I think it's done.

t-imamichi commented:
When I generate html, I notice that `OptimizationResult` is not included in html. Should I change `qiskit.optimization.algorithms.__init__` as follows?

```
Results
==========

.. autosummary::
   :toctree: ../stubs/
   :nosignatures:

   OptimizationResult
   ADMMOptimizationResult
   MinimumEigenOptimizerResult
```

t-imamichi commented:
I checked the classes in qiskit.optimization.algorithms module.
Which should we include in the docstring of `__init__.py`?

- admm_optimizer.py
  - ADMMOptimizer(OptimizationAlgorithm)
  - ADMMOptimizationResult(OptimizationResult)
  - ADMMParameter
  - ADMMState
- cobyla_optimizer.py
  - CobylaOptimizer(MultiStartOptimizer)
- cplex_optimizer.py
  - CplexOptimizer(OptimizationAlgorithm)
- grover_optimizer.py
  - GroverOptimizer(OptimizationAlgorithm)
  - GroverOptimizationResults
- minimum_eigen_optimizer.py
  - MinimumEigenOptimizer(OptimizationAlgorithm)
  - MinimumEigenOptimizerResult(OptimizationResult)
- multistart_optimizer.py
  - MultiStartOptimizer(OptimizationAlgorithm)
- optimization_algorithm.py
  - OptimizationAlgorithm
  - OptimizationResult
  - OptimizationResultStatus
- recursive_minimum_eigen_optimizer.py
  - RecursiveMinimumEigenOptimizer(OptimizationAlgorithm)
- slsqp_optimizer.py
  - SlsqpOptimizer(MultiStartOptimizer)

woodsp-ibm commented:
@t-imamichi I believe all those classes are exposed via the pubic API whether directly or as part of some result e.g. ADMMState is exposed via a getter of ADMMOptimizationResult. Hence all of them should be there I think. Don't forget to tweak the `GroverOptimizationResults` name to drop the 's' from the end so its consistent with the other result objects

t-imamichi commented:
I updated the docstring of `__init__.py` and added a reno text. Could you take a look at it? @woodsp-ibm

woodsp-ibm commented:
Can we get docstrings on the constructors - most importantly on any that we have in the that list now which are the public API. For instance OptimizationResult, ADMMOptimzationResult to name a couple. What also seems a little odd is that ADMMOptimizationResult constructor has the same set of params as the base class, plus some, but the set that it has from the base class are in a different order. 
Also GroverOptimizationResult does not extend the OptimizationResult base class despite the GroverOptimizer extending the OptimzationProblem. Rather GroverOptimizer returns a the base class Optimization result or which the GroverOptimizationResult is something in the raw_results field. The naming is just a little confusing here thats all leading one to mistakening think that GroverOptimizationResult is like the others. Maybe since this is just part of the raw results we can rename it to GroverOptimizerRawResult or something that does not make it appear to be an OptimzationResult sub-class like the others

woodsp-ibm commented on qiskit/optimization/algorithms/__init__.py:
```diff
@@ -28,34 +28,43 @@
    :nosignatures:
 
    OptimizationAlgorithm
+   OptimizationResult
```
Since this class is more than an abstract base class, but is a result object that is often returned by an Optimizer that does not sub-class it to add extra fields, it may perhaps be better in the other list directly in with Algorithms and results.

t-imamichi replied:
It makes sense. I move it.

t-imamichi commented:
The constructor of ADMMOptimzationResult does not have a docstring. I will add it. I also rename GroverOptimizationResult with GroverOptimizationRawResult.

t-imamichi commented:
- Added docstrings of ADMMOptimziationResult and MinimumEigenOptimizerResult.
- Renamed GroverOptimizationResult -> GroverOptimizationRawResult
- Moved OptimizationResult to 'Algorithms and results' in the main docstring.

a-matsuo commented:
I just notice that `RecursiveMinimumEigenOptimizer.solve()` returns `OptimizationResult` whose `raw_results` is a tuple while other optimizers return some the corresponding result class whose `raw_results` is a dictionary. It might not be important, and we can ignore it.

woodsp-ibm commented:
@a-matsuo I had made a related comment earlier (copied from below from above since its getting hard to navigate things now in this PR) - the consensus seemed to be that this aspect would be looked at in another PR #1120 related to raw results
```
 woodsp-ibm 7 days ago Member
Should there be some structure to raw_results, as in rather than Any would Dict[str, Any] be preferable?
I see this for instance the GroverOptimizer passes in a dictionary there. Not sure for other cases. 
At least the string field names give some clue to the content which might not otherwise be there if
its Any directly.

adekusar-drl 7 days ago Member
I'd suggest to keep it as is now and return to raw_result in #1120, if you don't mind.

t-imamichi 7 days ago Author Member
I'd like to keep raw_results just Any in this PR and discuss details of raw_results in #1120 too. Is it OK?
```

a-matsuo commented:
@woodsp-ibm oops. Thank you. I had not read all the comments and I wrote it right away after I noticed.

t-imamichi commented:
I'm working on resolving the conflicts. Stay tuned.

t-imamichi commented:
Thank you for your cooperation!
